x/operator/keeper/slash.go

```diff
@@ -33,215 +33,101 @@ func GetSlashIDForDogfood(infraction stakingtypes.Infraction, infractionHeight i
-	for assetID, state := range historyOperatorAssets {
-		slashAmount := parameter.SlashProportion.MulInt(state.TotalAmount).TruncateInt()
-		slashAssets[assetID] = slashAmount
+	executionInfo := &types.SlashExecutionInfo{
+		SlashProportion:    newSlashProportion,
+		SlashValue:         slashUSDValue,
+		SlashUndelegations: make([]*types.SlashFromUndelegation, 0),
+		SlashAssetsPool:    make([]*types.SlashFromAssetsPool, 0),
 	}
-
 	// slash from the unbonding stakers
 	if parameter.SlashEventHeight < ctx.BlockHeight() {
 		// get the undelegations that are submitted after the slash.
 		opFunc := func(undelegation *delegationtype.UndelegationRecord) error {
-			totalSlashAmount, ok := slashAssets[undelegation.AssetID]
-			if ok {
-				slashAmount := parameter.SlashProportion.MulInt(undelegation.Amount).TruncateInt()
-				remainingSlashAmount, err := SlashFromUndelegation(undelegation, totalSlashAmount, slashAmount)
-				if err != nil {
-					return err
-				}
-				// slashing from the operator isn't needed if the remainingSlashAmount isn't positive
-				if !remainingSlashAmount.IsPositive() {
-					delete(slashAssets, undelegation.AssetID)
-				} else {
-					slashAssets[undelegation.AssetID] = remainingSlashAmount
-				}
+			slashFromUndelegation, err := SlashFromUndelegation(undelegation, newSlashProportion)
+			if err != nil {
+				return err
+			}
+			if slashFromUndelegation != nil {
+				executionInfo.SlashUndelegations = append(executionInfo.SlashUndelegations, slashFromUndelegation)
 			}
 			return nil
 		}
 		// #nosec G701
 		heightFilter := uint64(parameter.SlashEventHeight)
 		err = k.delegationKeeper.IterateUndelegationsByOperator(ctx, parameter.Operator.String(), &heightFilter, true, opFunc)
 		if err != nil {
-			return err
+			return nil, err
 		}
 	}
 
-	// slash the remaining from the assets pool of the operator
-	for assetID, slashAmount := range slashAssets {
-		operatorAsset, err := k.assetsKeeper.GetOperatorSpecifiedAssetInfo(ctx, parameter.Operator, assetID)
-		if err != nil {
-			return err
-		}
-		isClearUselessShare := false
-		if slashAmount.GTE(operatorAsset.TotalAmount) {
-			slashAmount = operatorAsset.TotalAmount
-			isClearUselessShare = true
-		}
+	// slash from the assets pool of the operator
+	opFuncToIterateAssets := func(assetID string, state *assetstype.OperatorAssetInfo) error {
+		slashAmount := newSlashProportion.MulInt(state.TotalAmount).TruncateInt()
+		remainingAmount := state.TotalAmount.Sub(slashAmount)
+		// todo: consider slash all assets if the remaining amount is too small,
+		// which can avoid the unbalance between share and amount
 
-		changeAmount := assetstype.DeltaOperatorSingleAsset{
-			TotalAmount: slashAmount.Neg(),
-		}
 		// all shares need to be cleared if the asset amount is slashed to zero,
 		// otherwise there will be a problem in updating the shares when handling
 		// the new delegations.
-		if isClearUselessShare {
+		if remainingAmount.IsZero() {
```

Comment:
Consider **very small residual amounts** to avoid calculation errors and imbalances

Reply:
Yeah, there has been a `todo` comment.